src/label_producer.py

```diff
@@ -0,0 +1,205 @@
+            length greater than 1 and are alphabetical characters
+        """
+        return [word for word in text if word not in stop_words and len(word) > 1 and word.isalpha()]
+    
+    def __tokenize(self, text, stop_words):
+        """
+        Tokenize the fiven text
+        
+        Args:
+            text: the text to tokenize
+            stop_words: a list of words to filter out
+            
+        Returns:
+            text after tokenization, conversion to lowercase and filtering of words
+            using the filterWords function
+        """
+        return self.__filter_words(nltk.word_tokenize(text), stop_words)
+    
+    def __term_counter(self, term, stop_words):
+        """
+        Count the occurrences of different words in the sentences that contain
+        all the words of the given term
+        
+        Args:
+            term: the term to query for
+            stop_words: a list of words to filter out
+            
+        Returns:
+            A counter of occurrences of all the unfiltered words that returned from
+            the given query
+        """
+        res = self.__get_pages(term)
+        result_words = []
+        for doc in res:
+            result_words += self.__tokenize(doc.lower(), stop_words)
+        counter = Counter(result_words)
+        return counter
+    
+    def calculate_most_probable_relations(self, terms, topn=20):
+        """
+        Get the most probable labels for the given list of terms
+        
+        Args:
+            terms: a list of 2-tuples which are terms to search labels for
+            topn: number of labels to return
+            
+        Returns:
+            A list of 'topn' labels which represent the relations between the words
+            in the terms sorted in descending order of probablity
+        """
+    
+        nltk.download('stopwords') 
+        nltk.download('brown') 
+        
+        stop_words = stopwords.words('english')
+        
+        frequencies_file_name = '../data/frequency.pickle'
+
+        term_str = list(map(lambda pair: pair[0] + " " + pair[1], terms))
+        
+        try:
+            print('Loading word frequencies dictionary...')
+            self.__TERM_FREQUENCIES_DICT = pickle.load(open(frequencies_file_name, 'rb'))
+        except:
+            print('Word frequencies dictionary doesn\'t exist; Creating it and saving it to file...')
+            self.__TERM_FREQUENCIES_DICT = FreqDist(word.lower() for word in brown.words())
+            os.makedirs(os.path.dirname(frequencies_file_name), exist_ok=True)
+            with open(frequencies_file_name, 'wb') as handle:
+                pickle.dump(self.__TERM_FREQUENCIES_DICT, handle, protocol = pickle.HIGHEST_PROTOCOL)
+                
+        print('Finished loading word frequencies dictionary')
+        
+        terms_set = set([word.lower() for term in term_str for word in term.split()])
```

Comment:
Since the _term_ is a result of join on _term_ in _terms_, we can skip the ```split``` and iterate on terms directly:
```[word.lower() for pair in terms for word in pair]```